umbral/umbral.py

```diff
@@ -120,98 +119,124 @@ def __bytes__(self):
 
         self.cfrags = {}
+        self._contents = None
 
-    @staticmethod
-    def from_bytes(data: bytes, curve: ec.EllipticCurve):
+    class NotValid(ValueError):
+        """
+        raised if the capusle does not pass verification.
+        """
+
+    @classmethod
+    def from_bytes(cls, data: bytes, curve: ec.EllipticCurve, is_reconstructed=False):
         """
         Instantiates a Capsule object from the serialized data.
         """
-        eph_e = Point.from_bytes(data[0:33], curve)
-        eph_v = Point.from_bytes(data[33:66], curve)
-        sig = BigNum.from_bytes(data[66:98], curve)
+        if is_reconstructed:
+            e_prime = Point.from_bytes(data[0:33], curve)
+            v_prime = Point.from_bytes(data[33:66], curve)
+            eph_ni = Point.from_bytes(data[66:99], curve)
 
-        return Capsule(eph_e, eph_v, sig)
+            return cls(e_prime=e_prime, v_prime=v_prime, noninteractive_point=eph_ni)
+        else:
+            eph_e = Point.from_bytes(data[0:33], curve)
+            eph_v = Point.from_bytes(data[33:66], curve)
+            sig = BigNum.from_bytes(data[66:98], curve)
 
-    def to_bytes(self):
+            return cls(eph_e, eph_v, sig)
+
+    @property
+    def contents(self):
+        return self._contents
+
+    def to_bytes(self, reconstructed_components=False):
         """
         Serialize the Capsule into a bytestring.
         """
-        eph_e = self.point_eph_e.to_bytes()
-        eph_v = self.point_eph_v.to_bytes()
-        sig = self.bn_sig.to_bytes()
+        if reconstructed_components:
+            eph_e = self._point_eph_e_prime
+            eph_v = self._point_eph_v_prime
+            point_noninter = self._point_noninteractive
+
+            return eph_e.to_bytes() + eph_v.to_bytes() + point_noninter.to_bytes()
+        else:
+            eph_e = self._point_eph_e.to_bytes()
+            eph_v = self._point_eph_v.to_bytes()
+            sig = self._bn_sig.to_bytes()
 
-        return eph_e + eph_v + sig
+            return eph_e + eph_v + sig
 
     def verify(self, params: UmbralParameters):
 
-        e = self.point_eph_e
-        v = self.point_eph_v
-        s = self.bn_sig
+        e = self._point_eph_e
+        v = self._point_eph_v
+        s = self._bn_sig
         h = hash_to_bn([e, v], params)
 
         return s * params.g == v + (h * e)
 
     def attach_cfrag(self, cfrag: CapsuleFrag):
         self.cfrags[cfrag.bn_kfrag_id] = cfrag
 
-    def reconstruct(self):
+    def open(self, pub_bob, priv_bob, pub_alice, force_reopen=False, pre=None):
+        # TODO: Raise an error here if Bob has gathered enough cFrags?  See #22.
+        if self.contents and not force_reopen:
+            newly_opened = True
+        else:
+            self._reconstruct()
+            if not pre:
+                pre = PRE(UmbralParameters())
+            self._contents = pre._decapsulate_reencrypted(pub_bob, priv_bob, pub_alice, self)
```

Comment:
Not sure if we want to store the decrypted key here. Ideally, it isn't stored except when it's needed due to how Python handles memory.

So, if we store it in the object and the object is just sorta kept around for any reason, so is the key. I like having the user declared variable be the only place where the key is stored. Not sure if @cygnusv feels the same way, though.

Reply:
> Ideally, it isn't stored except when it's needed due to how Python handles memory.

To which part of python memory handling are you referring, specifically?

> So, if we store it in the object and the object is just sorta kept around for any reason, so is the key. I like having the user declared variable be the only place where the key is stored. 

For most attacks I can think of, this is a false sense of security and not a genuine security measure. 

You don't *really* know when garbage collection is occurring in either case, although yes, it's probably true in most cases that a class instance with a attribute will persist as a reference longer than a simple name.  

You'll find that "accept it and move on" is a popular attitude on this topic, and that even solutions which purport to scrub object in memory rely on ctypes, are implementation specific, and still end up being vulnerable to an attacker who can read process memory.

However, if this vector is your concern - and thus your threat model includes access to process memory - then we need a wholesale re-evaluation of a lot of the logic of this project.  I think that, for now, we need proceed to launch with an understanding that we do not provide robustness against a compromised system.

I understand that in this age of VPS's and Spectres that it's tempting to make something that is perfectly secure against the apparent power of an attacker who can work your own hardware against you, but I think that's outside the scope of this project.

I *do*, however, like what I understand as your general thought process on the matter - for example, having a separate `KeyStore` process, with narrower entry and exit points, and having `Bob` store a `Capsule` with `m-1` `CFrags` attached.

So, first: let's work through which part of python memory handling concerns you - this is an area over which I have a reasonable thorough understanding.  Which aspect of the way python handles memory do you want to discuss?

Some resources that might help you think about this:

* https://security.stackexchange.com/questions/12332/where-to-store-a-key-for-encryption
* https://stackoverflow.com/questions/982682/mark-data-as-sensitive-in-python/983525#983525 (again, ctypes and implementation specific stuff here!)
* https://stackoverflow.com/questions/728164/securely-erasing-password-in-memory-python

Reply:
I'm not speaking on Python specifically. It's just generally considered a good idea to keep plaintext key material out of memory as long as possible.

Considering that the reconstructed capsule is there, it should be possible for the user to provide their private key whenever they need to actually access it.

This keeps the plaintext key out of the memory as much as possible and leaves it up to the user to maintain that themselves. My ideal method would be to not cache the key in plaintext, but to cache it as ciphertext -- if we have to.

Since that is a different engineering challenge and out of the scope of Umbral itself, I don't think it should be implemented now or even here.

Reply:
Yeah, I suppose that makes sense.